index.js

```diff
@@ -136,61 +136,104 @@ function RpcAgent(opts = {}) {
 }
 
 function rpc(request, callback) {
-  const requestSerialized = JSON.stringify(request);
-  const auth = new Buffer(`${this.user}:${this.pass}`).toString('base64');
-  const options = {
-    host: this.host,
-    port: this.port,
-    path: '/',
-    method: 'POST',
-  };
-  let err = null;
-  const req = this.prot.request(options, (res) => {
-    let buf = '';
-    res.on('data', (data) => {
-      buf += data;
+  return new Promise((resolve, reject) => {
+    const requestSerialized = JSON.stringify(request);
+    const auth = new Buffer(`${this.user}:${this.pass}`).toString('base64');
+    const options = {
+      host: this.host,
+      port: this.port,
+      path: '/',
+      method: 'POST',
+    };
+    let err = null;
+    const req = this.prot.request(options, (res) => {
+      let buf = '';
+      res.on('data', (data) => {
+        buf += data;
+      });
+      res.on('end', () => {
+        if (res.statusCode === 401) {
+          if(this.calb) {
+            return callback(new Error('bitcoin JSON-RPC connection rejected: 401 unauthorized'));
+          }
+          reject(new Error('bitcoin JSON-RPC connection rejected: 401 unauthorized'));
+          return
+        }
+        if (res.statusCode === 403) {
+          if(this.calb) {
+            return callback(new Error('bitcoin JSON-RPC connection rejected: 403 forbidden'));
+          }
+          reject(new Error('bitcoin JSON-RPC connection rejected: 403 forbidden'));
+          return
+        }
+        if (err) {
+          if(this.calb) {
+            return callback(err);
+          }
+          reject(err);
+          return
+        }
+        let bufDeserialized;
+        try {
+          bufDeserialized = JSON.parse(buf);
+        } catch (e) {
+          if(this.calb) {
+            return callback(e);
+          }
+          reject(e);
+          return
+        }
+        if(this.calb) {
+          return callback(bufDeserialized.error, bufDeserialized);
+        }
+        if(bufDeserialized.error) {
+            reject(bufDeserialized.error)
+            return
+        }
+        resolve(bufDeserialized.result);
+      });
     });
-    res.on('end', () => {
-      if (res.statusCode === 401) {
-        return callback(new Error('bitcoin JSON-RPC connection rejected: 401 unauthorized'));
-      }
-      if (res.statusCode === 403) {
-        return callback(new Error('bitcoin JSON-RPC connection rejected: 403 forbidden'));
-      }
-      if (err) {
-        return callback(err);
+    req.on('error', (e) => {
+      err = new Error(`Could not connect to bitcoin via RPC at \
+            host: ${this.host} port: ${this.port} Error: ${e.message}`);
+      if (this.calb) {
+        callback(err);
       }
-      let bufDeserialized;
-      try {
-        bufDeserialized = JSON.parse(buf);
-      } catch (e) {
-        return callback(e);
-      }
-      return callback(bufDeserialized.error, bufDeserialized);
+      reject(err);
```

Comment:
When `calb` is set, both `callback()` and `reject()` will be called. Use `else` for the `!this.calb` case, or return in the above if block?